packages/math/src/SphericalHarmonics3.ts

```diff
@@ -0,0 +1,268 @@
+  private static _basisFunction = [
+    0.282095, //  1/2 * Math.sqrt(1 / PI)
+
+    -0.488603, // -1/2 * Math.sqrt(3 / PI)
+    0.488603, //  1/2 * Math.sqrt(3 / PI)
+    -0.488603, // -1/2 * Math.sqrt(3 / PI)
+
+    1.092548, //  1/2 * Math.sqrt(15 / PI)
+    -1.092548, // -1/2 * Math.sqrt(15 / PI)
+    0.315392, //  1/4 * Math.sqrt(5 / PI)
+    -1.092548, // -1/2 * Math.sqrt(15 / PI)
+    0.546274 //  1/4 * Math.sqrt(15 / PI)
+  ];
+
+  private static _convolutionKernel = [
+    3.141593, //  PI
+    2.094395, // (2 * PI) / 3,
+    0.785398 //   PI / 4
+  ];
+
+  private _coefficients: Float32Array = new Float32Array(27);
+
+  /**
+   * Convert radiance to irradiance with the A_l which is convoluted by the cosine lobe and pre-scale the basis function.
+   * @remarks
+   * Reference equation [4,5,6,7,8,9] from https://graphics.stanford.edu/papers/envmap/envmap.pdf
+   *
+   * @param out - The array
+   * @returns Pre-scaled array
+   */
+  convertRadianceToIrradiance(out: Float32Array): Float32Array {
+    const kernel = SphericalHarmonics3._convolutionKernel;
+    const basis = SphericalHarmonics3._basisFunction;
+    const src = this._coefficients;
+
+    /**
+     * 1.  L -> E
+     * 2.  E * basis
+     */
+
+    // l0
+    out[0] = src[0] * kernel[0] * basis[0];
+    out[1] = src[1] * kernel[0] * basis[0];
+    out[2] = src[2] * kernel[0] * basis[0];
+
+    // l1
+    out[3] = src[3] * kernel[1] * basis[1];
+    out[4] = src[4] * kernel[1] * basis[1];
+    out[5] = src[5] * kernel[1] * basis[1];
+    out[6] = src[6] * kernel[1] * basis[2];
+    out[7] = src[7] * kernel[1] * basis[2];
+    out[8] = src[8] * kernel[1] * basis[2];
+    out[9] = src[9] * kernel[1] * basis[3];
+    out[10] = src[10] * kernel[1] * basis[3];
+    out[11] = src[11] * kernel[1] * basis[3];
+
+    // l2
+    out[12] = src[12] * kernel[2] * basis[4];
+    out[13] = src[13] * kernel[2] * basis[4];
+    out[14] = src[14] * kernel[2] * basis[4];
+    out[15] = src[15] * kernel[2] * basis[5];
+    out[16] = src[16] * kernel[2] * basis[5];
+    out[17] = src[17] * kernel[2] * basis[5];
+    out[18] = src[18] * kernel[2] * basis[6];
+    out[19] = src[19] * kernel[2] * basis[6];
+    out[20] = src[20] * kernel[2] * basis[6];
+    out[21] = src[21] * kernel[2] * basis[7];
+    out[22] = src[22] * kernel[2] * basis[7];
+    out[23] = src[23] * kernel[2] * basis[7];
+    out[24] = src[24] * kernel[2] * basis[8];
+    out[25] = src[25] * kernel[2] * basis[8];
+    out[26] = src[26] * kernel[2] * basis[8];
+
+    return out;
+  }
+
+  /**
+   * Add radiance to the SH3 in specified direction.
+   * @remarks
+   * Implements `EvalSHBasis` from [Projection from Cube maps] in http://www.ppsloan.org/publications/StupidSH36.pdf.
+   *
+   * @param color - Radiance color
+   * @param direction - Radiance direction
+   * @param solidAngle - Radiance solid angle, dA / (r^2)
+   */
+  addRadiance(color: Color, direction: Vector3, solidAngle: number): void {
```

Comment:
The meaning of the parameter is direction light, but your function name is 'addRadiance'.
I think addDirectionLight is better and the third param use 'internsity' is better for debelopers.

Reply:
+1